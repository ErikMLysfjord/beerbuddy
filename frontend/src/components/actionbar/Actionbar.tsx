import { Select, Input } from "antd";
import styles from "./Actionbar.module.css";
import FilterButton from "../filter-button/FilterButton";
import { useContext } from "react";
import { FilterContext } from "../../context/FilterContext";
import useWindowDimensions from "../../utils/useWindowDimensions";
import SortingButton from "../sorting-button/SortingButton";

/**
 * The actionbar component. Contains the sorting and search functionality.
 * @param fetchMore - function that is called when the apply filters button is clicked
 * @returns - The actionbar component.
 */
const Actionbar = ({
  fetchMore,
}: {
  fetchMore: (reset?: boolean, noFilters?: boolean) => Promise<void>;
}) => {
  const { setSearchString, setSorting, sorting } = useContext(FilterContext);
  const { width } = useWindowDimensions();

  return (
    <section className={styles.wrapper}>
      {width > 768 && (
        <label className={styles.sortLabel}>
          Sort by
          <Select
            defaultValue={sorting}
            title="Sort by"
            style={{ width: 200 }}
            options={[
              { value: "top", label: "Most popular" },
              { value: "low", label: "Least popular" },
              { value: "atoz", label: "A-Z" },
              { value: "ztoa", label: "Z-A" },
            ]}
            onChange={(value) => setSorting(value)}
          />
        </label>
      )}
      <label className={styles.searchLabel}>
        Search
        <Input.Search
          onSearch={(value) => setSearchString(value.toLowerCase())}
        />
      </label>
      <FilterButton fetchMore={fetchMore} />
      {width <= 768 && (
        <SortingButton
          items={[
            { key: "top", label: "Most popular" },
            { key: "low", label: "Least popular" },
            { key: "atoz", label: "A-Z" },
            { key: "ztoa", label: "Z-A" },
          ]}
        />
      )}
    </section>
  );
};
export default Actionbar;
