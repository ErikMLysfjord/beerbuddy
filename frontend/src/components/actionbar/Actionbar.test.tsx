import { describe, it, expect } from "vitest";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Actionbar from "./Actionbar";
import { act } from "react-dom/test-utils";

describe("Actionbar", () => {
  it("renders correctly", () => {
    const { container } = render(<Actionbar />);
    expect(container).toMatchSnapshot();
  });

  it("renders with correct dropdown", () => {
    const { getByRole } = render(<Actionbar />);
    expect(getByRole("combobox")).toBeInTheDocument();
    expect(screen.queryByText("Ascending")).not.toBeInTheDocument();
    expect(screen.queryByText("Descending")).not.toBeInTheDocument();
  });

  it("renders option after clicking dropdown", async () => {
    render(<Actionbar />);

    const dropdown = screen.getByRole("combobox");

    act(() => {
      fireEvent.mouseDown(dropdown);
    });

    await waitFor(() => {
      expect(screen.getByText("Ascending")).toBeInTheDocument();
      expect(screen.getByText("Descending")).toBeInTheDocument();
    });
  });

  it("renders with correct searchbar", () => {
    const { getByRole } = render(<Actionbar />);
    expect(getByRole("textbox")).toBeInTheDocument();
  });
});
