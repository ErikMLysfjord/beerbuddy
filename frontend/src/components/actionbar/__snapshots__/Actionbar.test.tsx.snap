// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Actionbar > renders correctly 1`] = `
<div>
  <section
    aria-label="Search for and sort beers"
    class="_wrapper_ebfc65"
  >
    <label
      class="_sortLabel_ebfc65"
    >
      Sort by
      <div
        class="ant-select css-dev-only-do-not-override-pr0fja ant-select-single ant-select-show-arrow"
        style="width: 200px;"
        title="Sort by"
      >
        <div
          class="ant-select-selector"
        >
          <span
            class="ant-select-selection-search"
          >
            <input
              aria-autocomplete="list"
              aria-controls="rc_select_TEST_OR_SSR_list"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-label="Search"
              aria-owns="rc_select_TEST_OR_SSR_list"
              autocomplete="off"
              class="ant-select-selection-search-input"
              id="rc_select_TEST_OR_SSR"
              readonly=""
              role="combobox"
              style="opacity: 0;"
              type="search"
              unselectable="on"
              value=""
            />
          </span>
          <span
            class="ant-select-selection-item"
            title="Sort by"
          >
            Most popular
          </span>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
    </label>
    <label
      class="_searchLabel_ebfc65"
    >
      Search
      <span
        class="ant-input-group-wrapper ant-input-search css-dev-only-do-not-override-pr0fja"
      >
        <span
          class="ant-input-wrapper ant-input-group css-dev-only-do-not-override-pr0fja"
        >
          <input
            class="ant-input css-dev-only-do-not-override-pr0fja"
            name="search"
            type="text"
            value=""
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-default ant-btn-icon-only ant-input-search-button"
              type="button"
            >
              <span
                class="ant-btn-icon"
              >
                <span
                  aria-label="search"
                  class="anticon anticon-search"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="search"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </span>
        </span>
      </span>
    </label>
    <div
      class="_filterButtonWrapper_1f5523"
    >
      <button
        aria-label="Filter button"
        class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
        type="button"
      >
        <span
          class="ant-btn-icon"
        >
          <span
            aria-label="filter"
            class="anticon anticon-filter"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="filter"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </section>
</div>
`;
